### ISSUE TITLE
exabgp 4.2.6 has duplicated process when reload with usr2

### ISSUE THREAD

**Issue by sincerywaing** [Exa-Networks/exabgp#966]
version: 4.2.6

how to reproduce:

run exabgp with process, then you’ll see one process running as defined after run directive.

then update the command used by run, send a usr2 to exabgp.

You’ll see both process (old run command and new run command) are running.

Start and stop the service could clear the old command process.

another question is what’s the difference between sending usr1 and usr2? if I want a graceful reload (don’t want bgp neighbor down or route withdraw), seems usr2 is preferred?

**Comment by thomas-mangin** [comment]
Thank you @sincerywaing - I will look into it and see why this happening.

**Comment by sincerywaing** [comment]
another question is what’s the difference between sending usr1 and usr2? if I want a graceful reload (don’t want bgp neighbor down or route withdraw), seems usr2 is preferred?

@thomas-mangin what’s your suggestion on this one?

**Comment by sincerywaing** [comment]
@thomas-mangin are you able to reproduce the issue?

**Comment by thomas-mangin** [comment]
@sincerywaing I have not been working on ExaBGP, but I will get to it.

**Comment by sincerywaing** [comment]
I feel this has something to do with process keeps ending the routes every time it completes a healthcheck. `reload_completed` can't be finished.

**Comment by sincerywaing** [comment]
it also withdraw the route even there is new process announcing the route.

**Comment by thomas-mangin** [comment]
same bug as #972

**Comment by sincerywaing** [comment]
@thomas-mangin tested on 4.2.7 still same issue. duplicated process. I think this has something to do with stdin issue with this version, as the stale process is waiting for stdin and stuck there.

**Comment by thomas-mangin** [comment]
what is the code of your process?

**Comment by thomas-mangin** [comment]
@sincerywaing 
[humouristic mode on]
My ego is simply too large to entertain the possibility of a bug in my code. Therefore I can only assume the issue is in this invisible code of yours .. 

If you want **free** support and help for a **free tool**, in my not so available free time (**no my time is NOT FREE**) ... please provide me with something which allows me to help you. Otherwise, I will have no other choice that to close this issue with "works for me" and invite you to contact your favourite vendors' $$$ TAC for help with an alternative product.
[humouristic mode off]

Can you please provide me with something which allows me to reproduce the problem you see, please? It is broken is not enough and does not allow me to help you.

**Comment by sincerywaing** [comment]
the ultimate fix is to remove `no-ack` and `sudo` from `mutually_exclusve_arguments` :)

**Comment by shaoberry** [comment]
I meet this issue in 4.2.7 as well and find it only happens when ack is enabled. 

**Symptom:**
```systemctl status exabgp4
● exabgp4.service - LSB: start and stop exabgp
   Loaded: loaded (/etc/rc.d/init.d/exabgp4; bad; vendor preset: disabled)
   Active: active (running) since Tue 2020-05-26 07:22:19 UTC; 4s ago
     Docs: man:systemd-sysv-generator(8)
  Process: 33794 ExecStop=/etc/rc.d/init.d/exabgp4 stop (code=exited, status=0/SUCCESS)
  Process: 13366 ExecReload=/etc/rc.d/init.d/exabgp4 reload (code=exited, status=0/SUCCESS)
  Process: 34940 ExecStart=/etc/rc.d/init.d/exabgp4 start (code=exited, status=0/SUCCESS)
    Tasks: 4
   Memory: 61.4M
   CGroup: /system.slice/exabgp4.service
           ├─34976 /usr/local/bin/python3.7 /usr/local/bin/exabgp -e /path/exabgp.env /path/ecmp.exabgp.cfg
           ├─34978 /usr/local/bin/python -m exabgp.application.healthcheck -F /path/announce-unicast-ip4.cfg
           ├─34979 /usr/local/bin/python -m exabgp.application.healthcheck -F /path/announce-ip4.cfg
           └─34980 /usr/local/bin/python3.7 /usr/local/bin/exabgp

systemctl reload exabgp4

systemctl status exabgp4
● exabgp4.service - LSB: start and stop exabgp
   Loaded: loaded (/etc/rc.d/init.d/exabgp4; bad; vendor preset: disabled)
   Active: active (running) since Tue 2020-05-26 07:22:19 UTC; 26s ago
     Docs: man:systemd-sysv-generator(8)
  Process: 33794 ExecStop=/etc/rc.d/init.d/exabgp4 stop (code=exited, status=0/SUCCESS)
  Process: 35459 ExecReload=/etc/rc.d/init.d/exabgp4 reload (code=exited, status=0/SUCCESS)
  Process: 34940 ExecStart=/etc/rc.d/init.d/exabgp4 start (code=exited, status=0/SUCCESS)
    Tasks: 8
   Memory: 83.2M
   CGroup: /system.slice/exabgp4.service
           ├─34976 /usr/local/bin/python3.7 /usr/local/bin/exabgp -e /path/exabgp.env /path/ecmp.exabgp.cfg
           ├─34978 /usr/local/bin/python -m exabgp.application.healthcheck -F /path/announce-unicast-ip4.cfg
           ├─34979 /usr/local/bin/python -m exabgp.application.healthcheck -F /path/announce-ip4.cfg
           ├─35476 /usr/local/bin/python -m exabgp.application.healthcheck -F /path/announce-unicast-ip4.cfg
           ├─35479 /usr/local/bin/python -m exabgp.application.healthcheck -F /path/announce-ip4.cfg
           └─35481 /usr/local/bin/python3.7 /usr/local/bin/exabgp
```
each reload will make a set of whole healthcheck processes.

**Analysis:**
the old processes are waiting from input from stdin.
```strace -p 34978,34979
strace: Process 34978 attached
strace: Process 34979 attached
[pid 34978] read(0,  <unfinished ...>
[pid 34979] read(0,```

The call stack shows its originated from sigterm_handler. It seems it finishes all cleaning work and just wait for an ack to exit.
 ``` File "/usr/local/lib/python3.7/site-packages/exabgp/application/healthcheck.py", line 604, in sigterm_handler
    exabgp(states.EXIT)
  File "/usr/local/lib/python3.7/site-packages/exabgp/application/healthcheck.py", line 525, in exabgp
    sys.stdin.readline()
```

**Root Cause:**
when SIGUSR2 is delivered, the main process sends SIGTERM to hc(healthcheck) processes and directly deletes hc processes altogether with their stdio pipes from its watchlist. However, the SIGTERM handler of hc sends out withdraw command and blockingly wait for an ack from main process which has already had no way to read/write those pipes.

**Workaround:**
disable ack for main process and healthcheck process, but I have to run it as root.

**Permanant Solution:**
Could you please kindly review it? [delete process from watchlist until it dies](https://github.com/Exa-Networks/exabgp/pull/981).

**Comment by sincerywaing** [comment]
above fix is in 4.2.9
